Deployable set 1

rahul-400 opened the pull request.

rahul-400 committed:
Addition of access token based authorisation

rahul-400 committed:
Exception handling addition

rahul-400 committed:
Correction of few apis

rahul-400 committed:
Commit

rahul-400 committed:
{\rtf1\ansi\ansicpg1252\cocoartf2511
\cocoatextscaling0\cocoaplatform0{\fonttbl\f0\fswiss\fcharset0 Helvetica;}
{\colortbl;\red255\green255\blue255;}
{\*\expandedcolortbl;;}
\paperw11900\paperh16840\margl1440\margr1440\vieww10800\viewh8400\viewkind0
\pard\tx566\tx1133\tx1700\tx2267\tx2834\tx3401\tx3968\tx4535\tx5102\tx5669\tx6236\tx6803\pardirnatural\partightenfactor0

\f0\fs24 \cf0 Things implemented in deployable set 1:\
1) Entity and dots for company, department list, department, employee, salary.\
2) Created controller and wrote Crud operations for all the above mentioned entity.\
3) Created a generic api that fetches the entire company details.}

kapilmangal1622 commented on src/main/java/com/assignment/application/RedisConfig.java:
```diff
@@ -1,40 +1,49 @@
-    public JedisConnectionFactory jedisConnectionFactory(){
-        RedisStandaloneConfiguration redisStandaloneConfiguration = new RedisStandaloneConfiguration("localhost",6379);
+    public JedisConnectionFactory jedisConnectionFactory() {
+        RedisStandaloneConfiguration redisStandaloneConfiguration = new RedisStandaloneConfiguration("localhost", 6379);
```
redis host name should be environment variable.

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -22,99 +27,77 @@
+    public ResponseEntity<CompanyDTO> addCompany(@RequestBody CompanyDTO companyDTO,
+                                                 @RequestHeader("access_token") String token) {
+
+        verifyUsers.authorizeUser(token, "company", "post");
```
declare these static strings where-ever used in StringConstant class.

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -22,99 +27,77 @@
+                                                           @RequestParam(defaultValue = "0") Integer pageNo,
+                                                           @RequestParam(defaultValue = "10") Integer pageSize) {
+        verifyUsers.authorizeUser(token, "company", "get");
+        Pageable pageable = PageRequest.of(pageNo, pageSize);
```
Take pageable in arguments. No need to initialize it seperately.
Change it where-ever used.

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -22,99 +27,77 @@
     private CompanyServiceI companyServiceI;
```
Change the class name to CompanyService

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -22,99 +27,77 @@
+        return new ResponseEntity<>(StringConstant.USER_VERIFIED, HttpStatus.OK);
+    }
+
+    public Company convertToEntity(CompanyDTO companyDTO) {
```
Move this function to util or service class.

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/DepartmentController.java:
```diff
@@ -19,101 +25,82 @@
+    @PostMapping(value = "/{company_id}/department")
+    public ResponseEntity<DepartmentDTO> addDepartment(@PathVariable("company_id") Long companyId,
+                                                       @RequestBody DepartmentDTO departmentDTO,
+                                                       @RequestHeader("access_token") String token) {
```
access_token , company_id must be declared in stringConstant class

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/DepartmentController.java:
```diff
@@ -19,101 +25,82 @@
     private DepartmentServiceI departmentServiceI;
```
Change the class name

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -22,99 +27,77 @@
+        verifyUsers.authorizeUser(token, "company", "post");
+        Company company = convertToEntity(companyDTO);
+        company = companyServiceI.createNewCompany(company);
+        companyDTO.setId(company.getId());
```
Call convertToDto function here.

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/DepartmentController.java:
```diff
@@ -19,101 +25,82 @@
+        return new ResponseEntity<>(StringConstant.DELETION_SUCCESSFUL, HttpStatus.OK);
+    }
+
+    public Department convertToEntity(DepartmentDTO departmentDTO) {
```
Move these function in util or service class so that these can be used anywhere

kapilmangal1622 commented on src/main/java/com/assignment/application/controller/DepartmentListController.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping(value = "/departmentlist")
```
rename it to department-list

kapilmangal1622 commented on src/main/java/com/assignment/application/dto/CompanyDTO.java:
```diff
@@ -0,0 +1,76 @@
+
+import java.io.Serializable;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
```
You can use @ Data annotation here. No need to write setters and getters.

kapilmangal1622 commented on src/main/java/com/assignment/application/entity/Company.java:
```diff
@@ -1,45 +1,77 @@
-@Table(name="company_info")
-public class Company {
+@Table(name = "company_info")
+public class Company implements Serializable {
```
declare created_at, created_by, updated_at, updated_by in another class and extend this class (inheritance).

kapilmangal1622 commented on src/main/java/com/assignment/application/entity/Company.java:
```diff
@@ -1,45 +1,77 @@
+
+    @Column(name = "created_at")
+    @JsonFormat(pattern = "yyyy-MM-dd")
+    private Date createdAt;
```
No need to set created_at, updated_at manually.
explore

kapilmangal1622 commented on src/main/java/com/assignment/application/entity/CompleteCompInfo.java:
```diff
@@ -1,40 +1,43 @@
 import java.io.Serializable;
 
 @Component
 public class CompleteCompInfo implements Serializable {
```
This class can be renamed to companyInfo. Is complete required?

kapilmangal1622 commented on src/main/java/com/assignment/application/kafka/KafkaConfig.java:
```diff
@@ -33,69 +35,82 @@
-        Map<String,Object> map = new HashMap<>();
+    public ProducerFactory<String, EmployeeInfoUpdate> producerFactoryEmpUpdate() {
+        Map<String, Object> map = new HashMap<>();
         map.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, StringConstant.HOSTNAME);
```
declare hostname in application-properties (not in StringConstant class). Host name would not be same on staging. Any configuration must be declared in application-x.properties

kapilmangal1622 commented on src/main/java/com/assignment/application/kafka/KafkaConfig.java:
```diff
@@ -33,69 +35,82 @@
     @Bean
-    public ProducerFactory<String, SalaryUpdate> producerFactoryEmpSalary(){
-        Map<String,Object> map = new HashMap<>();
+    public ProducerFactory<String, SalaryUpdate> producerFactoryEmpSalary() {
```
why kafka configuration have been declared twice??

rahul-400 replied:
Because I have different model for different topic. Employee model class for employee information topic and salary update class for salary update topic.

kapilmangal1622 replied:
Declare those Configuration in a function and call the function as required

kapilmangal1622 commented on src/main/java/com/assignment/application/kafka/KafkaConfig.java:
```diff
@@ -33,69 +35,82 @@
-    public ConsumerFactory<String, KafkaEmployee> consumerFactory(){
-        Map<String,Object> map = new HashMap<>();
+    public ConsumerFactory<String, KafkaEmployee> consumerFactory() {
+        Map<String, Object> map = new HashMap<>();
```
declare kafkaProducer and kafkaconsumer configuration in seperate classes.

kapilmangal1622 commented on src/main/java/com/assignment/application/repo/SalaryRepo.java:
```diff
@@ -1,19 +1,30 @@
-    @Query("Select sal from Salary sal where sal.companyId=?1 and sal.departmentId=?2")
-    List<Salary> salaryListCompDept(Long compId,Long deptId);
+    @Query(value = "Select sal from Salary sal where sal.companyId=?1 and sal.departmentId=?2", nativeQuery = true)
+    List<Salary> salaryListCompDept(Long compId, Long deptId);
```
No need to write query here. change function name to getSalaryByCompanyIdAndDepartmentId

kapilmangal1622 commented on src/main/java/com/assignment/application/service/CachingInfo.java:
```diff
@@ -25,51 +29,87 @@
+        if (!employeeInfoUpdate.getPermanentAddress().isEmpty()) {
+            employee.setPermanentAddress(employeeInfoUpdate.getPermanentAddress());
         }
+        employee.setUpdatedAt(new Date());
```
no need to set UpdatedAt manually. declare a annotation on updated_at which will set the date

rahul-400 committed:
{\rtf1\ansi\ansicpg1252\cocoartf2511
\cocoatextscaling0\cocoaplatform0{\fonttbl\f0\fswiss\fcharset0 Helvetica;}
{\colortbl;\red255\green255\blue255;}
{\*\expandedcolortbl;;}
\paperw11900\paperh16840\margl1440\margr1440\vieww10800\viewh8400\viewkind0
\pard\tx566\tx1133\tx1700\tx2267\tx2834\tx3401\tx3968\tx4535\tx5102\tx5669\tx6236\tx6803\pardirnatural\partightenfactor0

\f0\fs24 \cf0 Things implemented in deployable set 1:\
1) Implemented triggers in database and did changes suggested(except few where I was getting some error)}

SuchiKarn commented on src/main/java/com/assignment/application/authenticator/VerifyUsers.java:
```diff
@@ -0,0 +1,40 @@
+    private RedisService redisService;
+
+    public int authorizeUser(String token, String url, String type) {
+        //        String cachedInfo = redisService.getKeyValue(StringConstant.ACCESS_TOKEN_REGEX + token);
+        //        if(cachedInfo.indexOf("roles: superadmin")==-1){
+        //            String[] info = token.split("-");
+        //            String[] urlSplit = url.split("/");
+        //            Company company = companyRepo.findById(Long.parseLong(info[info.length-1])).orElse(null);
+        //            String companyName = company.getName();
+        //            String employeeId = companyName.toLowerCase() + "_" + info[0];
+        //            if(urlSplit.length==3 && (urlSplit[2].equalsIgnoreCase("salary") || urlSplit[2].equalsIgnoreCase("update-employee-info")) &&
+        //                    (type.equalsIgnoreCase("GET") || type.equalsIgnoreCase("PATCH"))){
+        //                if(urlSplit[1].equalsIgnoreCase(employeeId) && info[info.length-1].equals(urlSplit[0])){
+        //                    return 1;
+        //                }
+        //                throw new UnauthorisedException("Not allowed to access");
+        //            }
+        //            throw new UnauthorisedException("Not allowed to access");
+        //        }
```
Remove commented codes. Add final and @NonNull for parameters passed in methods. Look for this in each file.

SuchiKarn commented on src/main/java/com/assignment/application/constants/StringConstant.java:
```diff
@@ -0,0 +1,39 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
Why @Component here?

SuchiKarn commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -1,120 +1,87 @@
-        } catch (Exception e) {
-            return new ResponseEntity<>(null, HttpStatus.BAD_REQUEST);
-        }
+    public ResponseEntity<CompanyDTO> addCompany(@RequestBody CompanyDTO companyDTO,
```
Use @Valid to validate incoming data.

SuchiKarn commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -1,120 +1,87 @@
+        verifyUsers.authorizeUser(token, "company", "post");
+        Company company = companyUtil.convertToEntity(companyDTO);
+        company = companyService.createNewCompany(company);
+        return new ResponseEntity<>(companyUtil.convertToDTO(company), HttpStatus.OK);
```
You can use mapstruct for converting DTO to Entity or Entity to DTO.

SuchiKarn commented on src/main/java/com/assignment/application/controller/CompanyController.java:
```diff
@@ -1,120 +1,87 @@
 
     @PatchMapping(value = "/{id}/company-update")
     public ResponseEntity<String> updateCompanyInfo(@PathVariable("id") Long id,
                                                     @RequestBody CompanyInfoUpdate companyInfoUpdate,
```
@Valid for object coming. @NonNull for path variable.

SuchiKarn commented on src/main/java/com/assignment/application/util/SalaryUtil.java:
```diff
@@ -0,0 +1,23 @@
+        return modelMapper.map(salary, SalaryDTO.class);
+    }
+
+}
```
Try using mapstruct. You won't need to create Util for each entity.

SuchiKarn commented on src/main/java/com/assignment/application/util/EmployeeUtil.java:
```diff
@@ -0,0 +1,29 @@
+package com.assignment.application.util;
+
+import com.assignment.application.dto.EmployeeDTO;
+import com.assignment.application.entity.Employee;
+import org.modelmapper.ModelMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDate;
+import java.util.Base64;
+
+@Component
+public class EmployeeUtil {
+
+    @Autowired
+    private ModelMapper modelMapper;
+
+    public Employee convertToEntity(EmployeeDTO employeeDTO) {
+        return modelMapper.map(employeeDTO, Employee.class);
+    }
+
+    public EmployeeDTO convertToDTO(Employee employee) {
+        EmployeeDTO employeeDTO = modelMapper.map(employee, EmployeeDTO.class);
+        String dob = new String(Base64.getDecoder().decode(employee.getDob()));
+        employeeDTO.setDob(LocalDate.parse(dob));
+        return employeeDTO;
+    }
+
+}
```
Don't create separate util for each entity. Look for mapstruct.

SuchiKarn commented on src/main/java/com/assignment/application/update/SalaryUpdate.java:
```diff
@@ -53,19 +58,19 @@ public void setDepartmentName(String departmentName) {
+    public Long getCompanyId() {
         return companyId;
     }
 
-    public void setCompanyId(long companyId) {
+    public void setCompanyId(Long companyId) {
         this.companyId = companyId;
     }
 
-    public long getValue() {
+    public Long getValue() {
         return value;
     }
 
-    public void setValue(long value) {
+    public void setValue(Long value) {
         this.value = value;
     }
```
Use @Data annotation instead of creating getters and setters for each field.

rahul-400 replied:
It creates a problem for the mapping I am using because of id field. So in few files I need to declare getters and setters.

SuchiKarn commented on src/main/java/com/assignment/application/service/SalaryServiceImpl.java:
```diff
@@ -93,32 +127,76 @@ public String updateSalary(Long companyId, SalaryUpdate salaryUpdate) {
 
-
+    @Override
+    public String updateSalaryOfEmployee(Long companyId, String employeeId, SalaryEmployeeUpdate salaryEmployeeUpdate) {
```
Standard practice is to make method of max 20 lines. Avoid big methods. Create private functions for particular actions which can be used in multiple other functions. Instead of using setField, use Builder. All fields value can be set at once using Builder.

SuchiKarn commented on src/main/java/com/assignment/application/service/SalaryServiceImpl.java:
```diff
@@ -40,47 +45,76 @@
 
+    //part of deployable set 3
     @Override
     public String updateSalary(Long companyId, SalaryUpdate salaryUpdate) {
```
Avoid creating methods bigger than 20 lines.

SuchiKarn commented on src/main/java/com/assignment/application/service/EmployeeServiceImpl.java:
```diff
@@ -27,59 +37,107 @@
     }
 
     @Override
     public String deleteEmployee(Long companyId, String employeeId) {
```
use final and @NonNull for parameters passed in methods.
